pkg/bitlpm/trie.go

```diff
@@ -0,0 +1,462 @@
+// compare prefix, and match length between the keys are equal to
+// the same value then the key is deleted from the trie.
+//
+// Note: Delete sets any prefixLen argument that exceeds the maximum
+// prefix allowed by the trie to the maximum prefix allowed by the
+// trie.
+func (t *trie[K, T]) Delete(prefixLen uint, k Key[K]) bool {
+	if k == nil {
+		return false
+	}
+	prefixLen = min(prefixLen, t.maxPrefix)
+
+	var (
+		grandParent, parent *node[K, T]
+		matchLen            uint
+		bitVal, prevBitVal  uint8
+	)
+
+	currentNode := t.root
+	for currentNode != nil {
+		// Find to what extent the current node matches with the
+		// delete-{prefix,key}.
+		matchLen = t.longestPrefixMatch(currentNode, prefixLen, k)
+		// The current-node does not match or it has the same
+		// prefix length (the only potential deletion in the
+		// trie).
+		if currentNode.prefixLen != matchLen ||
+			currentNode.prefixLen == prefixLen {
+			break
+		}
+		prevBitVal = bitVal
+		bitVal = k.BitValueAt(currentNode.prefixLen)
+		// We preserve the grandParent in order
+		// to prune intermediate nodes when they
+		// are no longer necessary.
+		grandParent = parent
+		parent = currentNode
+		currentNode = currentNode.children[bitVal]
+	}
+	// Not found, or the current-node does not match
+	// the delete-prefix exactly, or the current-node
+	// does not match the delete-{prefix,key} lookup,
+	// or the current-node is intermediate.
+	if currentNode == nil ||
+		currentNode.prefixLen != prefixLen ||
+		currentNode.prefixLen != matchLen ||
+		currentNode.intermediate {
+		return false
+	}
+	t.entries--
+
+	// If this node has two children, we need to keep it as an intermediate
+	// node because we cannot migrate both children up the trie.
+	if currentNode.children[0] != nil && currentNode.children[1] != nil {
+		currentNode.intermediate = true
+		return true
+	}
+
+	// If the parent of the current-node to be deleted is an
+	// intermediate-node and the current-node has no children
+	// then the parent (intermediate) node can be deleted and
+	// its other child promoted up the trie.
+	if parent != nil && parent.intermediate &&
+		currentNode.children[0] == nil && currentNode.children[1] == nil {
+		var saveNode *node[K, T]
+		if k.BitValueAt(parent.prefixLen) == 0 {
+			saveNode = parent.children[1]
+		} else {
+			saveNode = parent.children[0]
+		}
+		parent.children[0] = nil
+		parent.children[1] = nil
+		if grandParent == nil {
+			t.root = saveNode
+		} else {
+			grandParent.children[prevBitVal] = saveNode
+		}
+		return true
+	}
+
+	if currentNode.children[0] != nil {
+		parent.children[bitVal] = currentNode.children[0]
+	} else if currentNode.children[1] != nil {
+		parent.children[bitVal] = currentNode.children[1]
+	} else {
+		if parent == nil {
+			t.root = nil
+		} else {
+			parent.children[bitVal] = nil
+		}
+	}
+	return true
+}
+
+func (t *trie[K, T]) Len() uint {
+	return t.entries
+}
+
+func (t *trie[K, T]) ForEach(f ForEachFunc[Key[K], T]) {
```

Comment:
What are the ordering guarantees of ForEach?

Note: the seekable Iterator type of go-iradix is pretty useful; you might consider adopting that kind of API. It would solve the sub-range iterator problem nicely.

Reply:
There are no guarantees, it is a naive full traversal (traversing down all of the 0 bits first).